### PR TITLE
[libc++][ranges] use `static operator()` for C++23 ranges

### PR DESCRIPTION
**Pull request by xiaoyang-sde**
## Abstract

This pull request converts the `operator()` of all CPOs and niebloids related to C++23 ranges to `static`.

## Motivation

In `libc++`, CPOs and niebloids are implemented as function objects. Currently, the `operator()` for such a function object is a `const`-qualified member function. This means that even if the function object is has no data members, an extra register is used to pass in the `this` pointer when calling `operator()`, unless the compiler can inline the function call. Declaraing `operator()` as `static` would optimize away the unnecessary `this` pointer passing for stateless function objects, since there is no object instance state that needs to be accessed.

## Reference

- [P1169R4: static `operator()`](https://wg21.link/P1169R4)

### CODE REVIEW

File: libcxx/include/__ranges/chunk_by_view.h [hunk 1]
```diff
@@ -205,17 +205,19 @@ namespace views {
...
 
   template <class _Pred>
     requires constructible_from<decay_t<_Pred>, _Pred>
-  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI constexpr auto operator()(_Pred&& __pred) const
-      noexcept(is_nothrow_constructible_v<decay_t<_Pred>, _Pred>) {
-    return __range_adaptor_closure_t(std::__bind_back(*this, std::forward<_Pred>(__pred)));
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr auto
+  operator()(_Pred&& __pred) noexcept(is_nothrow_constructible_v<decay_t<_Pred>, _Pred>) {
+    constexpr auto __self = __fn{};
+    return __range_adaptor_closure_t(std::__bind_back(__self, std::forward<_Pred>(__pred)));
```
Comment (xiaoyang-sde) [inline libcxx/include/__ranges/chunk_by_view.h:220]: - `std::__bind_back(*this, std::forward<_Pred>(__pred))` won't compile because `this` is not available in a `static` function.
- `std::__bind_back(operator(), std::forward<_Pred>(__pred))` won't compile because the compiler can't determine which overload of `operator()` to use.

Please let me know if there's a better solution!

**Comment by xiaoyang-sde** [comment]
Once this pull request is merged, it's worth considering backporting this feature to those related to C++20 ranges, given that both Clang and GCC have implemented language extensions that support `static operator()` in C++20.

**Review by mordante — COMMENTED** [review]
Can you explain in the commit message why this change in beneficial?

File: libcxx/include/__ranges/chunk_by_view.h (hunk 1 above)
Comment (frederick-vs-ja) [reply libcxx/include/__ranges/chunk_by_view.h:220]: > The solution here is to initialize the function object `__fn` as a `constexpr` variable, which will be evaluated at compile time, and pass it to `std::__bind_back`. Please let me know if there's a better solution!

This looks good to me. [P2714R1](https://wg21.link/p2714r1) (adopted for C++26) exactly adds such things.

Comment (frederick-vs-ja) [reply libcxx/include/__ranges/chunk_by_view.h:220]: > The solution here is to initialize the function object `__fn` as a `constexpr` variable, which will be evaluated at compile time, and pass it to `std::__bind_back`. Please let me know if there's a better solution!

Perhaps we can "pass" the functor as a template argument in the same way as [P2714R1](https://wg21.link/p2714r1) (adopted for C++26).

Comment (mordante) [reply libcxx/include/__ranges/chunk_by_view.h:220]: I wonder whether we should change this place. We need a `*this` pointer so what is the benefit of this change?

File: libcxx/include/__ranges/zip_view.h [hunk 2]
```diff
@@ -489,12 +489,14 @@ namespace views {
...
-  _LIBCPP_HIDE_FROM_ABI constexpr auto operator()() const noexcept { return empty_view<tuple<>>{}; }
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr auto operator()() noexcept {
+    return empty_view<tuple<>>{};
+  }
 
   template <class... _Ranges>
-  _LIBCPP_HIDE_FROM_ABI constexpr auto operator()(_Ranges&&... __rs) const
-      noexcept(noexcept(zip_view<all_t<_Ranges&&>...>(std::forward<_Ranges>(__rs)...)))
-          -> decltype(zip_view<all_t<_Ranges&&>...>(std::forward<_Ranges>(__rs)...)) {
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI static constexpr auto
```
Comment (mordante) [inline libcxx/include/__ranges/zip_view.h:497]:
```suggestion
   _LIBCPP_HIDE_FROM_ABI static constexpr auto
```
When adding `_LIBCPP_NODISCARD_EXT` we need to test this property. I'm happy to approve this change + tests in a separate patch, but it should not be a drive-by in this one.

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for the updated commit message!
In general I like this patch, but a few comments.

File: libcxx/include/__ranges/chunk_by_view.h (hunk 1 above)
Comment (xiaoyang-sde) [reply libcxx/include/__ranges/chunk_by_view.h:220]: > I wonder whether we should change this place. We need a `*this` pointer so what is the benefit of this change?

This change allows the first overload of `operator()` to be `static`. We just need to figure out how to pass `__fn` to `std::__bind_back` here.

Comment (mordante) [reply libcxx/include/__ranges/chunk_by_view.h:220]: We can't this is a C++23 feature and P2714 is a C++26 feature.

**Review by mordante — COMMENTED** [review]
> Once this pull request is merged, it's worth considering backporting this feature to those related to C++20 ranges, given that both Clang and GCC have implemented language extensions that support `static operator()` in C++20. I'm willing to submit a pull request if that's a good idea.

I'm strongly against that idea. We agreed to allowed a small set of extensions to be used in older language versions to make the maintenance between language versions easier. We explicitly did not allow to use all extensions.

File: libcxx/include/__ranges/chunk_by_view.h (hunk 1 above)
Comment (xiaoyang-sde) [reply libcxx/include/__ranges/chunk_by_view.h:220]: @philnik777 @frederick-vs-ja What do you think?

Comment (frederick-vs-ja) [reply libcxx/include/__ranges/chunk_by_view.h:220]: > If P2714R1 is implemented

> We can't this is a C++23 feature and P2714 is a C++26 feature.

Oh, but IIUC we don't need to be waiting for P2714R1 being implemented. `__bind_back` is not a standard future and it seems OK to make it equivalent to C++26 `std::bind_back` in C++20 mode.

Comment (mordante) [reply libcxx/include/__ranges/chunk_by_view.h:220]: The  `__bind_back` implementation detail is used C++20 so that "runs" into the same issue. I really want to avoid this change in this PR. We can make a separate PR where we add a `__bind_back` for C++23 which does that. However in that case I want to see a measurable performance improvement before adding that.

Comment (xiaoyang-sde) [reply libcxx/include/__ranges/chunk_by_view.h:220]: > The `__bind_back` implementation detail is used C++20 so that "runs" into the same issue. I really want to avoid this change in this PR. We can make a separate PR where we add a `__bind_back` for C++23 which does that. However in that case I want to see a measurable performance improvement before adding that.

Sounds good to me. Thanks for reviewing this PR!

**Comment by philnik777** [comment]
> > Once this pull request is merged, it's worth considering backporting this feature to those related to C++20 ranges, given that both Clang and GCC have implemented language extensions that support `static operator()` in C++20. I'm willing to submit a pull request if that's a good idea.
> 
> I'm strongly against that idea. We agreed to allowed a small set of extensions to be used in older language versions to make the maintenance between language versions easier.

That was ever part of the reasoning. It's to reduce maintenance burden in general and possibly improve the QoI. IMO this clearly falls under "improve the QoI".

> We explicitly did not allow to use all extensions.

Yes. But that doesn't prevent us from allowing more useful extensions, such as this one.

**Comment by mordante** [comment]
> > > Once this pull request is merged, it's worth considering backporting this feature to those related to C++20 ranges, given that both Clang and GCC have implemented language extensions that support `static operator()` in C++20. I'm willing to submit a pull request if that's a good idea.
> > 
> > 
> > I'm strongly against that idea. We agreed to allowed a small set of extensions to be used in older language versions to make the maintenance between language versions easier.
> 
> That was ever part of the reasoning. It's to reduce maintenance burden in general and possibly improve the QoI. IMO this clearly falls under "improve the QoI".

I strongly disagree with this statement, especially the "clearly" part. I have not seen any numbers proving this change has a measurable benefit on code size or performance. Without these I don't know whether this will be a measurable improvement and when it is not how much.

> > We explicitly did not allow to use all extensions.
> 
> Yes. But that doesn't prevent us from allowing more useful extensions, such as this one.

It does not per se, but this is what we discussed and agreed upon. We didn't want to blanket allow all possible extensions.

**Comment by philnik777** [comment]
> I strongly disagree with this statement, especially the "clearly" part. I have not seen any numbers proving this change has a measurable benefit on code size or performance. Without these I don't know whether this will be a measurable improvement and when it is not how much.

Then feel free to look at the motivation for introducing the feature, which is referenced in the commit message from this very PR.

File: libcxx/include/__ranges/chunk_by_view.h (hunk 1 above)
Comment (xiaoyang-sde) [reply libcxx/include/__ranges/chunk_by_view.h]: The changes on this file have been reverted.

**Review by mordante — APPROVED** [review]
Thanks! LGTM!

**Comment by xiaoyang-sde** [comment]
> Thanks! LGTM!

Thanks for approving this PR! Could you please merge it once all CI checks have been passed?
